# Metis Final Project Part V: Concatenating all dataframes containing NYT article content into one dataframe that will be used for NLP. 

In [1]:
import pickle
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests

### The below code opens the dataframes that were stored in pickle files from previous sections of the project using pickle.load. 

### Further, the head() and .shape functions are called to observe that the dataframe has the correct dimensions and looks aesthetically 'OK'. 

In [3]:
text_dfs8_0to1000 = pickle.load( open('text_dfs8_0to1000','rb'))

In [4]:
text_dfs8_0to1000.head(2)

index          date                                                url  \
0      0                https://www.nytimes.com/2001/01/16/nyregion/c-...   
1      0  July 9, 2018  https://www.nytimes.com/2018/07/09/opinion/tru...   

                                             article    url_date  
0  A chart of financial indicators at the bottom ...  2001-01-16  
1  According to early indications, recent U.S. ec...  2018-07-09

In [5]:
text_dfs8_0to1000.shape

(1000, 5)

In [6]:
text_dfs8_1001to2000 = pickle.load( open('text_dfs8_1001to2000','rb'))

In [7]:
text_dfs8_1001to2000.head(2)

index          date                                                url  \
0      0                https://www.nytimes.com/2001/01/16/nyregion/c-...   
1      0  July 9, 2018  https://www.nytimes.com/2018/07/09/opinion/tru...   

                                             article    url_date  
0  A chart of financial indicators at the bottom ...  2001-01-16  
1  According to early indications, recent U.S. ec...  2018-07-09

In [8]:
text_dfs8_1001to2000.shape

(1000, 5)

In [9]:
text_dfs8_2001to3000 = pickle.load( open('text_dfs8_2001to3000','rb'))

In [10]:
text_dfs8_2001to3000.head(2)

index                                               date  \
0      0  JULY 10, 2011JULY 10, 2011SEP 5SEP 1AUG 28AUG ...   
1      0                                       JAN. 5, 2009   

                                                 url  \
0  https://www.nytimes.com/2011/07/11/opinion/11d...   
1  https://www.nytimes.com/2009/01/06/business/06...   

                                             article    url_date  
0  The Republican Party’s strategy in the debt-ce...  2011-07-11  
1  The Securities and Exchange Commission needs t...  2009-01-06

In [11]:
text_dfs8_2001to3000.shape

(999, 5)

In [13]:
text_dfs8_3001to4000 = pickle.load( open('text_dfs8_3001to4000_testing','rb'))

In [14]:
text_dfs8_3001to4000.head(2)

index          date                                                url  \
0      0   May 2, 2017  https://www.nytimes.com/2017/05/02/business/au...   
1      0  Feb. 1, 2015  https://www.nytimes.com/2015/02/02/business/ec...   

                                             article    url_date  
0  For seven years, the steadily expanding auto i...  2017-05-02  
1  Here are some of the top business stories to p...  2015-02-02

In [15]:
text_dfs8_3001to4000.shape

(999, 5)

In [16]:
text_dfs8_4001to5000_testing = pickle.load( open('text_dfs8_4001to5000_testing','rb'))

In [40]:
text_dfs8_4001to5000_testing.columns = ['index','date', 'url','article']

In [41]:
text_dfs8_4001to5000_testing.head(2)

index         date                                                url  \
0      0  MAY 4, 2009  https://www.nytimes.com/2009/05/05/opinion/05t...   
1      0               https://www.nytimes.com/2001/11/07/opinion/whi...   

                                             article  
0  There is something wrong with a system where s...  
1  Yesterday the Federal Reserve cut interest rat...

In [20]:
text_dfs8_5001to7000_testing = pickle.load( open('text_dfs8_5001to7000_testing','rb'))

In [23]:
print(text_dfs8_5001to7000_testing.shape)
text_dfs8_5001to7000_testing.head(2)

(1999, 4)


index           date                                                url  \
0      0  JUNE 21, 2006  https://www.nytimes.com/2006/06/21/business/21...   
1      0  Oct. 29, 2014  https://www.nytimes.com/2014/10/30/upshot/the-...   

                                             article  
0  Wall Street pared an earlier advance to end mi...  
1  The most important thing to understand about t...

In [24]:
text_dfs8_7001to10000_testing = pickle.load( open('text_dfs8_7001to10000_testing','rb'))

In [25]:
print(text_dfs8_7001to10000_testing.shape)
text_dfs8_7001to10000_testing.head(2)

(2999, 4)


index                                    date  \
0      0  NOV. 1, 2006MARCH 9, 2015MARCH 9, 2015   
1      0                            OCT. 7, 2008   

                                                 url  \
0  https://www.nytimes.com/2006/11/01/business/01...   
1  https://www.nytimes.com/2008/10/08/business/08...   

                                             article  
0  A couple of years ago, Robert E. Rubin — émine...  
1  DETROIT — The few consumers who are buying new...

In [28]:
text_dfs8_10001to13000_testing = pickle.load( open('text_dfs8_10001to13000_testing','rb'))

In [29]:
print(text_dfs8_10001to13000_testing.shape)
text_dfs8_10001to13000_testing.head(2)

(2999, 4)


index                                         date  \
0      0  DEC. 15, 2012AUG 29AUG 19JUL 26JUL 24JUL 15   
1      0                                 OCT. 2, 2011   

                                                 url  \
0  https://www.nytimes.com/2012/12/16/opinion/sun...   
1  https://www.nytimes.com/2011/10/03/business/ec...   

                                             article  
0  When President-elect Barack Obama selected Tim...  
1  ECONOMIC REPORTS This week’s data will include...

In [30]:
text_dfs8_13001to14000_testing = pickle.load( open('text_dfs8_13001to14000_testing','rb'))

In [31]:
print(text_dfs8_13001to14000_testing.shape)
text_dfs8_13001to14000_testing.head(2)

(999, 4)


index           date                                                url  \
0      0  JUNE 16, 2013  https://www.nytimes.com/2013/06/17/business/ec...   
1      0  NOV. 23, 2009  https://www.nytimes.com/2009/11/23/us/politics...   

                                             article  
0  ECONOMIC REPORTS Data to be released this week...  
1  President Obama’s scramble for a health care o...

In [32]:
text_dfs8_14001to15000_testing = pickle.load( open('text_dfs8_14001to15000_testing','rb'))

In [33]:
print(text_dfs8_14001to15000_testing.shape)
text_dfs8_14001to15000_testing.head(2)

(999, 4)


index                        date  \
0      0                JAN. 7, 2011   
1      0  November 25, 2012 10:43 pm   

                                                 url  \
0  https://www.nytimes.com/2011/01/08/business/ec...   
1  https://economix.blogs.nytimes.com/2012/11/25/...   

                                             article  
0  WASHINGTON — President Obama went to a busy wi...  
1  Every year, the government spends more than $1...

In [34]:
text_dfs8_15001to16000_testing = pickle.load( open('text_dfs8_15001to16000_testing','rb'))

In [35]:
print(text_dfs8_15001to16000_testing.shape)
text_dfs8_15001to16000_testing.head(2)

(999, 4)


index                       date  \
0      0      March 7, 2013 4:33 pm   
1      0  January 28, 2014 11:00 am   

                                                 url  \
0  https://query.nytimes.com/gst/fullpage.html?re...   
1  https://economix.blogs.nytimes.com/2014/01/28/...   

                                             article  
0  9:00 p.m. | UpdatedFour years after the financ...  
1  Phillip Swagel is a professor at the School of...

In [36]:
text_dfs8_16001to17000_testing = pickle.load( open('text_dfs8_16001to17000_testing','rb'))

In [37]:
print(text_dfs8_16001to17000_testing.shape)
text_dfs8_16001to17000_testing.head(2)

(998, 4)


index          date                                                url  \
0      0                https://www.nytimes.com/2002/01/12/business/wh...   
1      0  FEB. 7, 2013  https://www.nytimes.com/2013/02/08/business/ec...   

                                             article  
0  Wholesale prices in the United States declined...  
1  WASHINGTON (Reuters) — The number of Americans...

In [38]:
text_dfs8_17001to19702_testing = pickle.load( open('text_dfs8_17001to19702_testing','rb'))

In [39]:
print(text_dfs8_17001to19702_testing.shape)
text_dfs8_17001to19702_testing.head(2)

(2701, 4)


index                       date  \
0      0     June 16, 2011 11:02 am   
1      0  December 15, 2014 1:03 pm   

                                                 url  \
0  https://dealbook.nytimes.com/2011/06/16/wall-s...   
1  https://boss.blogs.nytimes.com/2014/12/15/toda...   

                                             article  
0  Wall Street is stepping up its attacks on new ...  
1  What’s affecting me, my clients and other sm...

In [42]:
frames = [text_dfs8_4001to5000_testing,text_dfs8_5001to7000_testing,text_dfs8_7001to10000_testing,text_dfs8_10001to13000_testing,text_dfs8_13001to14000_testing,text_dfs8_14001to15000_testing,text_dfs8_15001to16000_testing,text_dfs8_16001to17000_testing,text_dfs8_17001to19702_testing]

### For the the first part of the NYT_API Scraping, the end results are concatenated using pd.concat(), resulting in a dataframe that contains article content from 15,692 articles. 

### As mentioned in previous sections of the project, this dataframe contains articles that are dated from 2000-2018. 

In [43]:
result1 = pd.concat(frames,axis=0)

In [45]:
result1.shape

(15692, 4)

## Extracting Date Fields

### Although the date field for any given article is available to be scraped, I found that it was more efficient to omit any date tags from the scraping. Rather, once the article URL was scraped, I wrote a simple function implementing regex logic to parse this URL and return only the date identified in the URL. 

### Once the date has been extracted, the apply function is used to generate a new column in the dataframe containing the correct date value.


In [46]:
import re, datetime
def regex_date(string):
    try:
        match = re.search('\d{4}/\d{2}/\d{2}',string)
        date = datetime.datetime.strptime(match.group(), '%Y/%m/%d').date()
        return date
    except:
        return None

In [47]:
result1['url_date'] =result1['url'].apply(regex_date)

In [48]:
result1.head(3)

index          date                                                url  \
0      0   MAY 4, 2009  https://www.nytimes.com/2009/05/05/opinion/05t...   
1      0                https://www.nytimes.com/2001/11/07/opinion/whi...   
2      0  Dec. 4, 2014  https://www.nytimes.com/2014/12/04/opinion/rog...   

                                             article    url_date  
0  There is something wrong with a system where s...  2009-05-05  
1  Yesterday the Federal Reserve cut interest rat...  2001-11-07  
2  New York has always been a pretty good baromet...  2014-12-04

In [49]:
frames2 = [text_dfs8_0to1000,text_dfs8_1001to2000,text_dfs8_3001to4000,result1]

In [50]:
result2_part1 =  pd.concat(frames2,axis=0)

In [51]:
result2_part1.head(3)

index                                               date  \
0      0                                                      
1      0                                       July 9, 2018   
2      0  JAN. 22, 2008JAN. 21, 2008JAN. 22, 2008JAN. 22...   

                                                 url  \
0  https://www.nytimes.com/2001/01/16/nyregion/c-...   
1  https://www.nytimes.com/2018/07/09/opinion/tru...   
2  https://www.nytimes.com/2008/01/22/business/wo...   

                                             article    url_date  
0  A chart of financial indicators at the bottom ...  2001-01-16  
1  According to early indications, recent U.S. ec...  2018-07-09  
2  The Federal Reserve, responding to an internat...  2008-01-22

### The resulting dataframe contains 18,691 articles. 

In [52]:
result2_part1.shape

(18691, 5)

### The 'date' and 'index' columns are subsequently dropped. Scraping the NYT_API for these article values generated inconsistent results. 

In [53]:
columns = ['index','date']
result2_part1.drop(columns, axis=1, inplace=True)

In [54]:
result2_part1.head(2)

url  \
0  https://www.nytimes.com/2001/01/16/nyregion/c-...   
1  https://www.nytimes.com/2018/07/09/opinion/tru...   

                                             article    url_date  
0  A chart of financial indicators at the bottom ...  2001-01-16  
1  According to early indications, recent U.S. ec...  2018-07-09

In [55]:
result2_part1['article'].replace('', np.nan, inplace=True)

In [56]:
result2_part1.shape

(18691, 3)

### Given a cleaner dataframe, nulls are then removed. 

In [57]:
result2_part1.dropna(subset=['article'], inplace=True)

In [58]:
result2_part1.shape

(17724, 3)

### The resulting dataframe contains 17,724 rows of article content. The code executed below looks at random samples of the dataframe to ensure that everything looks 'OK'.

In [59]:
result2_part1.sample(50,random_state=42)

url  \
2145  https://www.nytimes.com/2004/05/23/business/ev...   
247   https://www.nytimes.com/2008/09/08/opinion/08k...   
1319  https://www.nytimes.com/2017/07/14/business/mu...   
95    https://www.nytimes.com/2011/10/03/opinion/hol...   
374   https://www.nytimes.com/2003/04/01/business/th...   
101   https://www.nytimes.com/2002/11/10/magazine/l-...   
1575  https://www.nytimes.com/2001/11/12/business/cn...   
29    https://www.nytimes.com/2003/08/29/business/ec...   
314   https://www.nytimes.com/2012/07/12/world/asia/...   
248   https://economix.blogs.nytimes.com/2013/01/04/...   
2852  https://www.nytimes.com/2008/12/03/business/ec...   
678   https://www.nytimes.com/2011/10/17/opinion/saf...   
691   https://www.nytimes.com/2003/04/03/business/a-...   
977   https://dealbook.nytimes.com/2010/03/07/consum...   
103   https://www.nytimes.com/2008/10/03/us/03immig....   
889   https://www.nytimes.com/2012/09/09/your-money/...   
1106  https://www.nytimes.com/2001/03/28/us/bush-say...   
147   https://www.nytimes.com/2015/03/31/business/ec...   
356   https://www.nytimes.com/2011/10/11/opinion/thi...   
2307  https://www.nytimes.com/2008/09/24/business/24...   
2192  https://www.nytimes.com/2008/08/29/business/29...   
499   https://www.nytimes.com/2012/09/26/business/ec...   
232   https://www.nytimes.com/2001/06/28/business/th...   
2246  https://www.nytimes.com/2006/09/22/us/politics...   
2613  https://www.nytimes.com/2008/12/06/business/ec...   
4     https://www.nytimes.com/2006/12/06/opinion/06w...   
1052  https://www.nytimes.com/2001/08/27/opinion/l-t...   
537   https://www.nytimes.com/2013/05/27/business/ec...   
949   https://www.nytimes.com/2010/05/30/business/30...   
2222  https://www.nytimes.com/2018/02/11/upshot/libe...   
2789  https://www.nytimes.com/2007/06/07/business/06...   
71    https://www.nytimes.com/2009/03/08/opinion/08r...   
684   https://www.nytimes.com/2011/10/10/business/ec...   
74    https://www.nytimes.com/2004/06/20/business/ma...   
1184  https://www.nytimes.com/2009/07/15/business/ec...   
1814  https://www.nytimes.com/2013/03/08/business/ec...   
1473  https://www.nytimes.com/2003/07/25/us/quick-bu...   
1934  https://www.nytimes.com/2015/12/07/opinion/hil...   
670   https://www.nytimes.com/2004/08/08/opinion/l-a...   
874   https://www.nytimes.com/2012/04/08/your-money/...   
1278  https://www.nytimes.com/2003/09/03/us/census-s...   
116   https://www.nytimes.com/2005/10/16/weekinrevie...   
392   https://www.nytimes.com/2002/02/10/business/po...   
413   https://boss.blogs.nytimes.com/2013/09/03/in-n...   
1765  https://dealbook.nytimes.com/2010/03/18/greens...   
364   https://www.nytimes.com/2014/10/11/upshot/a-fa...   
930   https://www.nytimes.com/2015/10/15/business/de...   
318   https://www.nytimes.com/2009/04/04/world/europ...   
590   https://economix.blogs.nytimes.com/2012/09/20/...   
50    https://www.nytimes.com/2010/09/12/opinion/12f...   

                                                article    url_date  
2145  LATELY, even more than usual, the word of the ...  2004-05-23  
247   Save the home lenders, save the world? If only...  2008-09-08  
1319  Who are you going to believe, the Fed or your ...  2017-07-14  
95    The dire state of the world economy reflects d...  2011-10-03  
374   DOLLAR FALLS. The dollar declined against the ...  2003-04-01  
101   When I first came across F. Scott Fitzgerald's...  2002-11-10  
1575  Think back to the distant past: last year.CNBC...  2001-11-12  
29    The economy grew at a revised annual rate of 3...  2003-08-29  
314   WASHINGTON — The White House on Wednesday form...  2012-07-12  
248   For jobs, the past four years have been a wash...  2013-01-04  
2852  WASHINGTONLast week, on the same day the Treas...  2008-12-03  
678   WashingtonTHERE’S a lot of talk in Washington ...  2011-10-17  
691   For many economists, the dollar's jagged yearl...  2003-04-03  
977   While much of the Congressional debate over co...  2010-

In [61]:
result2_part1.dtypes

url         object
article     object
url_date    object
dtype: object

### Having generated a clean dataframe from the first half of the overall articles scraped, the process is then repeated for the second half of the overall articles scraped. These articles date from 1982 to 2000.  

### As shown earlier, the files are opened and examined using the pickle.load() and head() functions. 

In [62]:
result2_part2 = pickle.load( open('nyt_result_part2','rb'))

In [63]:
result2_part2.head(4)

url  \
0  https://www.nytimes.com/1992/03/07/opinion/for...   
1  https://www.nytimes.com/1982/02/22/us/required...   
2  https://www.nytimes.com/1984/05/13/us/unions-e...   
3  https://www.nytimes.com/1996/01/24/business/in...   

                                             article    url_date  
0  There can hardly be doubt that in the next 6, ...  1992-03-07  
1  Attorney General William French Smith, speakin...  1982-02-22  
2  The American Federation of Labor and Congress ...  1984-05-13  
3  Personal incomes edged up just two-tenths of 1...  1996-01-24

In [64]:
result2_part2.shape

(10465, 3)

In [65]:
frames3 = [result2_part1,result2_part2]

### The articles from the second half of the scraping are then concatenated to the articles from the first half of the scraping. 

### The resulting dataframe is 28,189 rows long. 

In [66]:
complete_df = pd.concat(frames3)

In [67]:
complete_df.shape

(28189, 3)

### The code executed below generates a random sampling of the data to ensure that it looks 'OK'. 

In [68]:
complete_df.sample(100,random_state=42)

url  \
87    https://www.nytimes.com/2003/02/15/business/in...   
564   https://www.nytimes.com/1993/01/22/us/upturn-c...   
1548  https://www.nytimes.com/2011/06/20/business/ec...   
1575  https://www.nytimes.com/1982/06/01/opinion/l-a...   
659   https://www.nytimes.com/2002/11/17/business/ec...   
1572  https://www.nytimes.com/2015/06/30/business/de...   
1446  https://www.nytimes.com/2018/06/20/business/ec...   
734   https://www.nytimes.com/2009/02/09/us/politics...   
24    https://www.nytimes.com/2015/03/01/magazine/in...   
711   https://www.nytimes.com/2013/05/11/business/ec...   
1649  https://www.nytimes.com/1986/10/06/business/pu...   
341   https://www.nytimes.com/2014/02/14/business/ec...   
967   https://www.nytimes.com/2008/01/16/business/16...   
757   https://www.nytimes.com/1991/12/22/opinion/in-...   
2972  https://www.nytimes.com/2001/09/18/business/in...   
173   https://www.nytimes.com/1984/12/08/business/bo...   
254   https://economix.blogs.nytimes.com/2011/07/08/...   
699   https://query.nytimes.com/gst/fullpage.html?re...   
1133  https://www.nytimes.com/2008/03/23/opinion/23k...   
789   https://www.nytimes.com/2012/11/07/opinion/aft...   
590   https://www.nytimes.com/2002/11/10/magazine/l-...   
1218  https://www.nytimes.com/1982/08/18/business/ho...   
446   https://www.nytimes.com/2001/11/15/business/re...   
2640  https://www.nytimes.com/2010/11/16/business/ec...   
1364  https://www.nytimes.com/2008/01/13/business/13...   
8     https://www.nytimes.com/1983/01/11/business/ma...   
12    https://www.nytimes.com/1987/05/16/us/producer...   
1681  https://www.nytimes.com/1994/08/11/business/bu...   
830   https://www.nytimes.com/1987/06/17/business/fa...   
1032  https://www.nytimes.com/1992/05/30/business/co...   
...                                                 ...   
1693  https://www.nytimes.com/1991/03/24/business/lo...   
286   https://www.nytimes.com/2010/02/21/opinion/21f...   
777   https://www.nytimes.com/2006/04/29/business/29...   
703   https://www.nytimes.com/2009/11/17/business/ec...   
627   https://www.nytimes.com/1998/10/18/opinion/l-i...   
524   https://www.nytimes.com/2003/08/18/opinion/l-c...   
2282  https://www.nytimes.com/2018/03/28/opinion/tru...   
179   https://www.nytimes.com/2014/08/06/upshot/luck...   
422   https://www.nytimes.com/2012/12/04/business/ec...   
2314  https://www.nytimes.com/2009/02/20/travel/esca...   
802   https://www.nytimes.com/1986/09/19/business/gn...   
523   https://boss.blogs.nytimes.com/2014/03/06/toda...   
2317  https://economix.blogs.nytimes.com/2013/10/02/...   
304   https://www.nytimes.com/1984/09/22/us/briefing...   
1239  https://www.nytimes.com/2013/11/09/business/ec...   
216   https://www.nytimes.com/1997/05/10/business/le...   
1584  https://www.nytimes.com/1983/02/22/business/us...   
1987  https://www.nytimes.com/1986/06/22/weekinrevie...   
386   https://economix.blogs.nytimes.com/2010/08/12/...   
1122  https://www.nytimes.com/2012/10/23/us/across-c...   
836   https://www.nytimes.com/1985/12/08/business/we...   
760   https://www.nytimes.com/2003/08/14/us/kentucky...   
218   https://www.nytimes.com/2012/11/03/us/politics...   
2011  https://www.nytimes.com/2009/04/04/business/ec...   
156   https://www.nytimes.com/2012/09/22/business/ec...   
978   https://www.nytimes.com/2009/03/11/opinion/11w...   
1441  https://www.nytimes.com/2005/05/20/opinion/20k...   
1276  https://www.nytimes.com/1983/03/06/realestate/...   
459   https://www.nytimes.com/1990/04/22/business/bu...   
294   https://www.nytimes.com/2004/06/04/opinion/l-t...   

                                                article    url_date  
87    Though consumers remain the economy's main eng...  2003-02-15  
564   With the notable exception of California, the ...  1993-01-22  
1548  Two years into a fitful recovery, unemployed A...  2011-06-20  
1575  To the Editor:There have been some recent mino...  1982-06-01  
659   NOW that the Republicans have regaine

### Using the isnull function, we can observe that joining the two dataframes resulted in 231 null values. Those values are subsequently dropped using the dropna() function. 

In [69]:
complete_df.isnull().values.sum()

231

In [71]:
complete_df.dropna(inplace=True)

In [72]:
complete_df.shape

(27958, 3)

### The resulting dataframe contains 27,958 rows of article content spanning 1982 to 2018. The dataframe is then sampled to make sure everything looks OK. 

### The file is then stored as a pickle file. 

In [75]:
complete_df.sample(200,random_state=420)

url  \
377   https://www.nytimes.com/2011/03/26/business/ec...   
390   https://www.nytimes.com/2018/06/01/business/ec...   
590   https://www.nytimes.com/1988/11/06/business/da...   
920   https://www.nytimes.com/2015/12/05/business/ec...   
910   https://dealbook.nytimes.com/2014/04/29/big-te...   
1994  https://www.nytimes.com/1982/06/04/nyregion/qu...   
1158  https://www.nytimes.com/1982/08/22/nyregion/30...   
112   https://www.nytimes.com/1991/10/18/business/bi...   
287   https://www.nytimes.com/2010/09/15/realestate/...   
314   https://www.nytimes.com/2012/07/12/world/asia/...   
117   https://www.nytimes.com/2008/01/31/opinion/31e...   
1833  https://www.nytimes.com/2011/09/27/opinion/bro...   
516   https://www.nytimes.com/2011/10/19/arts/music/...   
551   https://www.nytimes.com/2011/05/14/opinion/l14...   
1308  https://www.nytimes.com/2007/01/05/business/05...   
13    https://www.nytimes.com/2013/03/02/business/ec...   
2517  https://www.nytimes.com/2001/04/28/business/ec...   
2723  https://www.nytimes.com/1991/12/25/business/ac...   
19    https://www.nytimes.com/2009/05/23/business/ec...   
2440  https://www.nytimes.com/1992/07/11/us/the-1992...   
2061  https://www.nytimes.com/2009/05/06/business/ec...   
296   https://economix.blogs.nytimes.com/2011/03/03/...   
288   https://www.nytimes.com/1993/02/14/business/ma...   
1306  https://www.nytimes.com/2012/03/30/business/da...   
582   https://www.nytimes.com/1985/08/05/business/pu...   
316   https://www.nytimes.com/1995/02/25/business/us...   
807   https://www.nytimes.com/2001/07/14/business/us...   
674   https://www.nytimes.com/1982/11/14/weekinrevie...   
1379  https://www.nytimes.com/1982/11/03/business/ec...   
6     https://www.nytimes.com/1999/11/17/business/sh...   
...                                                 ...   
2001  https://www.nytimes.com/2008/12/14/magazine/14...   
787   https://www.nytimes.com/2011/12/24/opinion/noc...   
1035  https://www.nytimes.com/1998/08/10/business/be...   
632   https://www.nytimes.com/1982/08/13/opinion/in-...   
1976  https://www.nytimes.com/2001/02/14/business/th...   
1148  https://www.nytimes.com/2008/12/18/business/wo...   
182   https://www.nytimes.com/1996/01/28/us/gramm-br...   
429   https://www.nytimes.com/2009/03/22/education/2...   
895   https://www.nytimes.com/1992/12/30/opinion/l-c...   
689   https://www.nytimes.com/1992/01/16/us/1992-cam...   
616   https://www.nytimes.com/2014/06/04/business/ec...   
1542  https://www.nytimes.com/1994/04/06/opinion/wal...   
633   https://www.nytimes.com/1983/01/20/business/st...   
2109  https://www.nytimes.com/2008/09/15/business/15...   
893   https://www.nytimes.com/2008/10/12/weekinrevie...   
410   https://www.nytimes.com/1988/08/13/us/call-to-...   
2830  https://www.nytimes.com/2005/10/14/business/14...   
1372  https://dealbook.nytimes.com/2010/04/13/dear-p...   
1193  https://www.nytimes.com/1989/11/19/business/bu...   
743   https://economix.blogs.nytimes.com/2010/08/03/...   
595   https://www.nytimes.com/2008/03/10/opinion/10k...   
2658  https://www.nytimes.com/2007/06/02/business/02...   
1509  https://www.nytimes.com/1983/11/20/business/ra...   
1253  https://economix.blogs.nytimes.com/2013/10/11/...   
222   https://www.nytimes.com/1986/10/01/business/pr...   
4     https://www.nytimes.com/1985/07/08/business/ma...   
232   https://www.nytimes.com/2009/10/08/us/politics...   
1523  https://www.nytimes.com/2004/06/13/business/su...   
1320  https://www.nytimes.com/2000/10/19/us/2000-cam...   
1110  https://www.nytimes.com/2000/02/23/opinion/l-e...   

                                                article    url_date  
377   The nation’s economy grew at a 3.1 percent ann...  2011-03-26  
390   Friday’s jobs report was a blockbuster: Job gr...  2018-06-01  
590                                         THE ECONOMY  1988-11-06  
920   Consider it a done deal.American employers exp...  2015-12-05  
910   Updated, 8:49 p.m. | It has long been

In [76]:
pickle.dump(complete_df, open( "complete_df_nyt_kojak", "wb" ) )